src/librustc_typeck/check/demand.rs

```diff
@@ -463,6 +463,57 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                     }
                 }
             }
+            _ if sp == expr.span => {
+                // If neither type is a reference, then check for `Deref` implementations by
+                // constructing a predicate to prove: `<T as Deref>::Output == U`
+                let deref_trait = self.tcx.lang_items().deref_trait().unwrap();
+                let item_def_id = self.tcx.associated_items(deref_trait).next().unwrap().def_id;
+                let predicate = ty::Predicate::Projection(ty::Binder::bind(ty::ProjectionPredicate {
+                    // `<T as Deref>::Output`
+                    projection_ty: ty::ProjectionTy {
+                        // `T`
+                        substs: self.tcx.mk_substs_trait(
+                            checked_ty,
+                            self.fresh_substs_for_item(sp, item_def_id),
+                        ),
+                        // `Deref::Output`
+                        item_def_id,
+                    },
+                    // `U`
+                    ty: expected,
+                }));
+                if self.infcx.predicate_may_hold(&traits::Obligation::new(
+                    self.misc(sp), self.param_env, predicate,
+                )) {
+                    // For a suggestion to make sense, the type would need to be `Copy`. Construct
+                    // a `T: Copy` predicate to prove.
+                    let copy_trait = self.tcx.lang_items().copy_trait().unwrap();
+                    let predicate = ty::Predicate::Trait(ty::Binder::bind(ty::TraitPredicate {
+                        // `T: Copy`
+                        trait_ref: ty::TraitRef::new(
+                            // `Copy`
+                            copy_trait,
+                            // `T`
+                            self.tcx.mk_substs_trait(
+                                expected,
+                                self.fresh_substs_for_item(sp, copy_trait),
+                            ),
+                        ),
+                    }));
+                    if self.infcx.predicate_may_hold(&traits::Obligation::new(
```

Comment:
copy checks can be done simpler (see the preexisting `*` suggestions above in https://github.com/rust-lang/rust/pull/60155/files#diff-124df07081a1185b5a4d40cdcf0cf4ffR443)

Reply:
Fixed.